samples/net/http_server/src/main.c

```diff
@@ -0,0 +1,673 @@
+ * not be used for anything else.
+ */
+#include "public_certificate.pem.h"
+#include "private_key.pem.h"
+#endif /* defined(CONFIG_NET_SOCKETS_SOCKOPT_TLS) */
+
+/* Macro called upon a fatal error, reboots the device. */
+#define FATAL_ERROR()					\
+	LOG_ERR("Fatal error! Rebooting the device.");	\
+	LOG_PANIC();					\
+	IF_ENABLED(CONFIG_REBOOT, (sys_reboot(0)))
+
+/* Zephyr NET management event callback structures. */
+static struct net_mgmt_event_callback l4_cb;
+static struct net_mgmt_event_callback conn_cb;
+
+struct http_req {
+	struct http_parser parser;
+	int socket;
+	bool received_all;
+	enum http_method method;
+	const char *url;
+	size_t url_len;
+	const char *body;
+	size_t body_len;
+};
+
+/* Forward declarations */
+static void process_tcp4(void);
+static void process_tcp6(void);
+
+/* Keep track of the current LED states. 0 = LED OFF, 1 = LED ON.
+ * Index 0 corresponds to LED1, index 1 to LED2.
+ */
+static uint8_t led_states[2];
+
+/* HTTP responses for demonstration */
+#define response_200 "HTTP/1.1 200 OK\r\n"
+#define response_400 "HTTP/1.1 400 Bad Request\r\n\r\n"
+#define response_403 "HTTP/1.1 403 Forbidden\r\n\r\n"
+#define response_404 "HTTP/1.1 404 Not Found\r\n\r\n"
+
+/* Processing threads for incoming connections */
+K_THREAD_STACK_ARRAY_DEFINE(tcp4_handler_stack, MAX_CLIENT_QUEUE, STACK_SIZE);
+static struct k_thread tcp4_handler_thread[MAX_CLIENT_QUEUE];
+static k_tid_t tcp4_handler_tid[MAX_CLIENT_QUEUE];
+K_THREAD_DEFINE(tcp4_thread_id, STACK_SIZE,
+		process_tcp4, NULL, NULL, NULL,
+		THREAD_PRIORITY, 0, -1);
+
+K_THREAD_STACK_ARRAY_DEFINE(tcp6_handler_stack, MAX_CLIENT_QUEUE, STACK_SIZE);
+static struct k_thread tcp6_handler_thread[MAX_CLIENT_QUEUE];
+static k_tid_t tcp6_handler_tid[MAX_CLIENT_QUEUE];
+K_THREAD_DEFINE(tcp6_thread_id, STACK_SIZE,
+		process_tcp6, NULL, NULL, NULL,
+		THREAD_PRIORITY, 0, -1);
+
+static K_SEM_DEFINE(network_connected_sem, 0, 1);
+static int tcp4_listen_sock;
+static int tcp4_accepted[MAX_CLIENT_QUEUE];
+static int tcp6_listen_sock;
+static int tcp6_accepted[MAX_CLIENT_QUEUE];
+static struct http_parser_settings parser_settings;
+
+static void l4_event_handler(struct net_mgmt_event_callback *cb,
+			     uint32_t event,
+			     struct net_if *iface)
+{
+	switch (event) {
+	case NET_EVENT_L4_CONNECTED:
+		LOG_INF("Network connected");
+
+		k_sem_give(&network_connected_sem);
+		break;
+	case NET_EVENT_L4_DISCONNECTED:
+		LOG_INF("Network disconnected");
+		break;
+	default:
+		/* Don't care */
+		return;
+	}
+}
+
+static void connectivity_event_handler(struct net_mgmt_event_callback *cb,
+				       uint32_t event,
+				       struct net_if *iface)
+{
+	if (event == NET_EVENT_CONN_IF_FATAL_ERROR) {
+		LOG_ERR("NET_EVENT_CONN_IF_FATAL_ERROR");
+		FATAL_ERROR();
+		return;
+	}
+}
+
+/* Update the LED states. Returns 0 if it was updated, otherwise -1. */
+static int led_update(uint8_t index, uint8_t state)
+{
+	int ret;
+
+	if (state <= 1) {
```

Comment:
Idea here was to keep the conversion from led id to index outside the LED update function. So that the LED update function only need to care about the index and state that it gets.

Reply:
fixed